File: website/api/detect.py
import os
import base64
import json
from io import BytesIO
from PIL import Image
import sys

# Add ultralytics to path if needed
try:
    from ultralytics import YOLO
except ImportError:
    pass

# Model cache
_model = None

def get_model():
    global _model
    if _model is None:
        try:
            from ultralytics import YOLO
            # Try to load the custom model
            model_path = os.path.join(os.path.dirname(__file__), '..', 'model', 'best.pt')
            if os.path.exists(model_path):
                _model = YOLO(model_path)
            else:
                # Fallback to yolov8n if custom model not found
                _model = YOLO('yolov8n.pt')
        except Exception as e:
            print(f"Error loading model: {e}")
            raise
    return _model

def handler(request):
    """Handle detection requests"""
    # Check request method
    if request.method == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type',
            }
        }
    
    if request.method != 'POST':
        return {
            'statusCode': 405,
            'body': json.dumps({'error': 'Method not allowed'}),
            'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'}
        }
    
    try:
        # Parse request body
        body = json.loads(request.body) if isinstance(request.body, str) else request.body
        
        if 'image' not in body:
            return {
                'statusCode': 400,
                'body': json.dumps({'error': 'No image provided'}),
                'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'}
            }
        
        # Decode base64 image
        image_data = body['image']
        if ',' in image_data:
            # Data URL format: data:image/jpeg;base64,...
            image_data = image_data.split(',')[1]
        
        image_bytes = base64.b64decode(image_data)
        image = Image.open(BytesIO(image_bytes))
        
        # Load model
        model = get_model()
        
        # Run detection
        results = model(image, conf=0.25)
        
        # Extract detections
        detections = []
        if results and len(results) > 0:
            result = results[0]
            if result.boxes is not None:
                for box in result.boxes:
                    detection = {
                        'class': model.names[int(box.cls[0])],
                        'confidence': float(box.conf[0]),
                        'bbox': [float(x) for x in box.xyxy[0].tolist()]
                    }
                    detections.append(detection)
        
        return {
            'statusCode': 200,
            'body': json.dumps(detections),
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            }
        }
    
    except Exception as e:
        print(f"Error: {e}")
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)}),
            'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'}
        }

File: website/api/detect.js
/**
 * Detection endpoint for IBCRS on Vercel
 * This proxies detection requests to an external YOLOv8 backend
 * 
 * Set the BACKEND_URL environment variable on Vercel to point to your Flask backend
 * Example: https://ibcrs-backend.railway.app/api/detect
 */
export default async function handler(req, res) {
  // Enable CORS
  res.setHeader('Access-Control-Allow-Credentials', 'true');
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST,OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    res.status(200).end();
    return;
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  // Get backend URL from environment variable
  const BACKEND_URL = process.env.YOLO_BACKEND_URL || process.env.BACKEND_URL;
  
  if (!BACKEND_URL) {
    return res.status(503).json({
      error: 'Backend not configured',
      message: 'Set YOLO_BACKEND_URL or BACKEND_URL environment variable to your Flask backend URL'
    });
  }

  try {
    const { image } = req.body;

    if (!image) {
      return res.status(400).json({ error: 'No image provided' });
    }

    // Proxy the request to the Flask backend
    const backendResponse = await fetch(`${BACKEND_URL}/api/detect`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ image })
    });

    if (!backendResponse.ok) {
      throw new Error(`Backend returned ${backendResponse.status}`);
    }

    const detections = await backendResponse.json();
    return res.status(200).json(detections);
  } catch (error) {
    console.error('Detection proxy error:', error);
    return res.status(500).json({
      error: 'Detection failed',
      message: error.message
    });
  }
}
